base/collection_gocb.go

```diff
@@ -40,11 +40,23 @@ var (
 func AsCollection(dataStore DataStore) (*Collection, error) {
 	collection, ok := dataStore.(*Collection)
 	if !ok {
-		return nil, errors.New("dataStore is not a *Collection")
+		return nil, fmt.Errorf("dataStore is not a %T", dataStore)
 	}
 	return collection, nil
 }
 
+// NewCollection creates a new collection object, caching some kv ops.
+func NewCollection(bucket *GocbV2Bucket, collection *gocb.Collection) (*Collection, error) {
+	c := &Collection{
+		Bucket:     bucket,
+		Collection: collection}
+	err := c.setCollectionID()
```

Comment:
This was the key piece I was missing, but the way this is still continuing to create collections on `NamedDataStore()` access means we're not actually caching this value beyond a single request.

As mentioned in Slack I think we may need to rethink how we store/fetch collections in general and have ideas about how to do this. But for now if we're able to set a collection ID for a collection it should be sufficient.